### ISSUE TITLE
Retry not working properly for all permutations of data-provider in Testng(7.0.0-beta)

### ISSUE THREAD

**Issue by sjethvani** [cbeust/testng#1946]
### TestNG Version
> Note: only the latest version is supported
Tetsng version : 7.0.0-beta1

### Expected behavior
In case if all permutations of data-provider fails for testcase (@test) then retry should happen for all of them. I executed same program with 6.10, 6.14.3 & 7.0.0-beta. Results are same for 6.10 & 6.14.3 , however they are different on 7.0.0-beta

### Actual behavior
Only first permutation of data-provider retries in 7.0.0-beta , while remaining don't . However in case of 6.10 & 6.14.3 all permutations of data-provider retries.

### Is the issue reproductible on runner?

- [ ] Shell
- [x] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [ ] IntelliJ
- [ ] NetBeans

### Test case sample
> Please, share the test case (as small as possible) which shows the issue

**RunnerMain.java (Main class running testng programmatically)**

```java
package com;
import org.testng.TestNG;

public class RunnerMain {

  public static void main(String[] args) {
    System.out.println("starting my testng programmatically");
    TestNG testNG = new TestNG();
    testNG.setTestClasses(new Class[] {Sample1.class, Sample2.class});
    testNG.addListener(new RetryAnnotationTransformation());
    testNG.run();
  }
}
```

**Sample1.java (containing some tests @test)**

```java
package com;

import org.testng.Assert;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

public class Sample1 {

  @DataProvider
  public Object[][] dp()
  {
    return new Object[][] { { "param1", "value1" }, { "param2", "value2" }};
  }
  
  @Test(dataProvider="dp")
  public void test1(String sUsername, String sPassword){
    System.out.println("-----under test1");
    System.out.println("uname is "+sUsername+"   --  password is "+sPassword);
    Assert.assertEquals(true, false,"NOt geting expeee");
    System.out.println("----test1 completed");
  }
  
  @Test(dataProvider="dp")
  public void test2(String sUsername, String sPassword){
    System.out.println("----under test2");
    System.out.println("uname is "+sUsername+"   --  password is "+sPassword);
    Assert.assertEquals(true, false,"ohh yeahh not getting ex");
    System.out.println("----test2 completed");
  }
  
}
```

**Sample2.java (containing some tests @test)**

```java
package com;

import org.testng.Assert;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

public class Sample2 {

  @DataProvider
  public Object[][] dp() {
    return new Object[][] { { "param1", "value1" }, { "param2", "value2" }};
  }
  
  @Test(dataProvider="dp")
  public void test3(String sUsername, String sPassword){
    System.out.println("-----under test3");
    System.out.println("uname is "+sUsername+"   --  password is "+sPassword);
    Assert.assertEquals(true, false,"wahhhhhh not gett");
    System.out.println("----test3 completed");
  }
  
  @Test(dataProvider="dp")
  public void test4(String sUsername, String sPassword){
    System.out.println("-----under test4");
    System.out.println("uname is "+sUsername+"   --  password is "+sPassword);
    Assert.assertEquals(true, false,"laaaaa not gett");
    System.out.println("----test4 completed");
  }
  
}
```

**RetryAnnotationTransformation.java (for applying Retries at annotation during run time)**

```java
package com;

import java.lang.reflect.Constructor;
import java.lang.reflect.Method;

import org.testng.IAnnotationTransformer;
import org.testng.IRetryAnalyzer;
import org.testng.annotations.ITestAnnotation;

/***
 * 
 * @author sjethvani
 *RetryListener testng listener runs whenever testcase fails and calls 'Retry' class's retry method
 */

public class RetryAnnotationTransformation implements IAnnotationTransformer{

  @Override
  public void transform(ITestAnnotation testannotation, Class testClass,Constructor testConstructor, Method testMethod) {
    IRetryAnalyzer retry = testannotation.getRetryAnalyzer();
    if (retry == null)  {
      testannotation.setRetryAnalyzer(RetryAnalyzer.class);
    }

  }
  
}
```

**RetryAnalyzer.java (implementation of IRetryAnalyzer interface)**

```java
package com;

import org.testng.IRetryAnalyzer;
import org.testng.ITestResult;

/***
 * 
 * @author sjethvani
 * Retry class to re-run failing test 
 *
 */
public class RetryAnalyzer implements IRetryAnalyzer {

  private int retryCount = 0;
  private int maxRetryCount = 1; //nuber of times we want to retry failing test

//Below method returns 'true' if the test method has to be retried else 'false' 
//and it takes the 'Result' as parameter of the test method that just ran
  public boolean retry(ITestResult result){
    System.out.println("retryCount is "+retryCount);
    System.out.println("maxRetryCount is "+maxRetryCount);
      if (retryCount < maxRetryCount) {
          if(result.getParameters().length>0)
          {
            System.out.println("Retrying Test [" + result.getName() +"] as its  status is [" + getResultStatusName(result.getStatus()) + "]. Parameters of above tests are "+result.getParameters()[0]);
          }
          else
          {
            System.out.println("Retrying test [" + result.getName() +"] as its status is [" + getResultStatusName(result.getStatus())+"]");
          }
          retryCount++;
          System.out.println("returning true for this test");
          return true;
      }
      System.out.println("returning false for this test "+result.getParameters()[0]);
      return false;
  }
  
  public String getResultStatusName(int status) {
    String resultName = null;
    if(status==1)
      resultName = "SUCCESS";
    if(status==2)
      resultName = "FAILURE";
    if(status==3)
      resultName = "SKIP";
  return resultName;
  }
}
```

If I run this prog with 6.10 or 6.14.3 then test count is as below 
```
===============================================
Command line suite
Total tests run: 16, Failures: 12, Skips: 4
===============================================
```

however with 7.0.0-beta1 , it is as below
```
===============================================
Command line suite
Total tests run: 12, Passes: 0, Failures: 8, Skips: 0, Retries: 4
===============================================
```

**Comment by sjethvani** [comment]
**7.0.0-beta**

```
starting my testng programmatically
-----under test1
uname is param1   --  password is value1
retryCount is 0
maxRetryCount is 1
Retrying Test [test1] as its  status is [FAILURE]. Parameters of above tests are param1
returning true for this test
-----under test1
uname is param1   --  password is value1
retryCount is 1
maxRetryCount is 1
returning false for this test param1
-----under test1
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2
----under test2
uname is param1   --  password is value1
retryCount is 0
maxRetryCount is 1
Retrying Test [test2] as its  status is [FAILURE]. Parameters of above tests are param1
returning true for this test
----under test2
uname is param1   --  password is value1
retryCount is 1
maxRetryCount is 1
returning false for this test param1
----under test2
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2
-----under test3
uname is param1   --  password is value1
retryCount is 0
maxRetryCount is 1
Retrying Test [test3] as its  status is [FAILURE]. Parameters of above tests are param1
returning true for this test
-----under test3
uname is param1   --  password is value1
retryCount is 1
maxRetryCount is 1
returning false for this test param1
-----under test3
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2
-----under test4
uname is param1   --  password is value1
retryCount is 0
maxRetryCount is 1
Retrying Test [test4] as its  status is [FAILURE]. Parameters of above tests are param1
returning true for this test
-----under test4
uname is param1   --  password is value1
retryCount is 1
maxRetryCount is 1
returning false for this test param1
-----under test4
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2

===============================================
Command line suite
Total tests run: 12, Passes: 0, Failures: 8, Skips: 0, Retries: 4
===============================================

```

**6.10**

```
starting my testng programmatically
[TestNG] Running:
  Command line suite

-----under test1
uname is param1   --  password is value1
retryCount is 0
maxRetryCount is 1
Retrying Test [test1] as its  status is [FAILURE]. Parameters of above tests are param1
returning true for this test
-----under test1
uname is param1   --  password is value1
retryCount is 1
maxRetryCount is 1
returning false for this test param1
-----under test1
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2
-----under test1
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2
----under test2
uname is param1   --  password is value1
retryCount is 0
maxRetryCount is 1
Retrying Test [test2] as its  status is [FAILURE]. Parameters of above tests are param1
returning true for this test
----under test2
uname is param1   --  password is value1
retryCount is 1
maxRetryCount is 1
returning false for this test param1
----under test2
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2
----under test2
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2
-----under test3
uname is param1   --  password is value1
retryCount is 0
maxRetryCount is 1
Retrying Test [test3] as its  status is [FAILURE]. Parameters of above tests are param1
returning true for this test
-----under test3
uname is param1   --  password is value1
retryCount is 1
maxRetryCount is 1
returning false for this test param1
-----under test3
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2
-----under test3
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2
-----under test4
uname is param1   --  password is value1
retryCount is 0
maxRetryCount is 1
Retrying Test [test4] as its  status is [FAILURE]. Parameters of above tests are param1
returning true for this test
-----under test4
uname is param1   --  password is value1
retryCount is 1
maxRetryCount is 1
returning false for this test param1
-----under test4
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2
-----under test4
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2

===============================================
Command line suite
Total tests run: 16, Failures: 12, Skips: 4
===============================================

```

**6.14.3**

```
starting my testng programmatically
-----under test1
uname is param1   --  password is value1
retryCount is 0
maxRetryCount is 1
Retrying Test [test1] as its  status is [FAILURE]. Parameters of above tests are param1
returning true for this test
-----under test1
uname is param1   --  password is value1
retryCount is 1
maxRetryCount is 1
returning false for this test param1
-----under test1
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2
-----under test1
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2
----under test2
uname is param1   --  password is value1
retryCount is 0
maxRetryCount is 1
Retrying Test [test2] as its  status is [FAILURE]. Parameters of above tests are param1
returning true for this test
----under test2
uname is param1   --  password is value1
retryCount is 1
maxRetryCount is 1
returning false for this test param1
----under test2
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2
----under test2
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2
-----under test3
uname is param1   --  password is value1
retryCount is 0
maxRetryCount is 1
Retrying Test [test3] as its  status is [FAILURE]. Parameters of above tests are param1
returning true for this test
-----under test3
uname is param1   --  password is value1
retryCount is 1
maxRetryCount is 1
returning false for this test param1
-----under test3
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2
-----under test3
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2
-----under test4
uname is param1   --  password is value1
retryCount is 0
maxRetryCount is 1
Retrying Test [test4] as its  status is [FAILURE]. Parameters of above tests are param1
returning true for this test
-----under test4
uname is param1   --  password is value1
retryCount is 1
maxRetryCount is 1
returning false for this test param1
-----under test4
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2
-----under test4
uname is param2   --  password is value2
retryCount is 1
maxRetryCount is 1
returning false for this test param2

===============================================
Command line suite
Total tests run: 16, Failures: 12, Skips: 4
===============================================
```

**Comment by krmahadevan** [comment]
@sjethvani - This is not an issue. TestNG has basically evolved when it comes to retrying of tests.
As part of fixing https://github.com/cbeust/testng/issues/1697 TestNG was enhanced to basically show retried tests separately.
When there's a retry analyzer involved, tests which got retried because the previous attempt failed.

"Skips" usually come into picture when an upstream test fails or if there is a configuration failure.

So what you are seeing is due to a behavior change in TestNG.

**Comment by sjethvani** [comment]
@krmahadevan well  , I am fine with testng's enhancement of retrying tests , all I was worried about was this 
`Only first permutation of data-provider retries in 7.0.0-beta , while remaining don't . However in case of 6.10 & 6.14.3 all permutations of data-provider retries`
From above attached logs , you can see that (param2,value2) is not being retried in Testng_7.0.0-beta , while it gets retried in previous versions of testng.

**Comment by krmahadevan** [comment]
@sjethvani - Thanks for being perseverant on this issue. I see the problem now. Working on getting this fixed. Will update shortly.

**Comment by krmahadevan** [comment]
@sjethvani - All the discrepancies should get resolved only my PR #1948 gets merged.

Please take a look at the [unit test](https://github.com/cbeust/testng/pull/1948/files#diff-a2689b8d38273c6aef88e6e10df5af05) in my PR #1948 and see if that addresses your concern :
